bap-vibes/src/linear_ssa.ml

```diff
@@ -67,77 +67,144 @@ let congruent (a : var) (b : var) : bool =
-    let new_sub_exp_3 = linearize_exp sub_exp_3 ~prefix in
+    let* new_sub_exp_1 = linearize_exp sub_exp_1 in
+    let* new_sub_exp_2 = linearize_exp sub_exp_2 in
+    let+ new_sub_exp_3 = linearize_exp sub_exp_3 in
     Bil.Store (new_sub_exp_1, new_sub_exp_2, new_sub_exp_3, endian, size)
   | Bil.UnOp (unop, sub_exp) -> 
-    let new_sub_exp = linearize_exp sub_exp ~prefix in
+    let+ new_sub_exp = linearize_exp sub_exp in
     Bil.UnOp (unop, new_sub_exp) 
   | Bil.BinOp (binop, sub_exp_1, sub_exp_2) ->
-    let new_sub_exp_1 = linearize_exp sub_exp_1 ~prefix in
-    let new_sub_exp_2 = linearize_exp sub_exp_2 ~prefix in
+    let* new_sub_exp_1 = linearize_exp sub_exp_1 in
+    let+ new_sub_exp_2 = linearize_exp sub_exp_2 in
     Bil.BinOp (binop, new_sub_exp_1, new_sub_exp_2)
   | Bil.Var v ->
-    let new_var = linearize v ~prefix in
+    let* new_var = linearize_var v in
+    let+ () = Linear.(update @@ Env.add_var new_var) in
     Bil.Var new_var
-  | Bil.Int _ -> exp
+  | Bil.Int _ -> Linear.return exp
   | Bil.Cast (cast, i, sub_exp) ->
-    let new_sub_exp = linearize_exp sub_exp ~prefix in
+    let+ new_sub_exp = linearize_exp sub_exp in
     Bil.Cast (cast, i, new_sub_exp)
   | Bil.Let (var, sub_exp_1, sub_exp_2) ->
-    let new_sub_exp_1 = linearize_exp sub_exp_1 ~prefix in
-    let new_sub_exp_2 = linearize_exp sub_exp_2 ~prefix in
+    let* new_sub_exp_1 = linearize_exp sub_exp_1 in
+    let+ new_sub_exp_2 = linearize_exp sub_exp_2 in
     Bil.Let (var, new_sub_exp_1, new_sub_exp_2)
   | Bil.Ite (sub_exp_1, sub_exp_2, sub_exp_3) ->
-    let new_sub_exp_1 = linearize_exp sub_exp_1 ~prefix in
-    let new_sub_exp_2 = linearize_exp sub_exp_2 ~prefix in
-    let new_sub_exp_3 = linearize_exp sub_exp_3 ~prefix in
+    let* new_sub_exp_1 = linearize_exp sub_exp_1 in
+    let* new_sub_exp_2 = linearize_exp sub_exp_2 in
+    let+ new_sub_exp_3 = linearize_exp sub_exp_3 in
     Bil.Ite (new_sub_exp_1, new_sub_exp_2, new_sub_exp_3)
   | Bil.Extract (i, j, sub_exp) ->
-    let new_sub_exp = linearize_exp sub_exp ~prefix in
+    let+ new_sub_exp = linearize_exp sub_exp in
     Bil.Extract (i, j, new_sub_exp) 
   | Bil.Concat (sub_exp_1, sub_exp_2) ->
-    let new_sub_exp_1 = linearize_exp sub_exp_1 ~prefix in
-    let new_sub_exp_2 = linearize_exp sub_exp_2 ~prefix in
+    let* new_sub_exp_1 = linearize_exp sub_exp_1 in
+    let+ new_sub_exp_2 = linearize_exp sub_exp_2 in
     Bil.Concat (new_sub_exp_1, new_sub_exp_2)
-  | Bil.Unknown (_, _) -> exp
+  | Bil.Unknown (_, _) -> Linear.return exp
 
-let linearize_phi ~(prefix : string) (phi : Phi.t) : Phi.t KB.t =
-  ignore prefix;
-  ignore phi;
-  Kb_error.(fail @@ Not_implemented "Linear_ssa.linearize_phi: unimplemented")
+let linearize_phi (phi : phi term) : phi term linear =
+  let lhs = Phi.lhs phi in
+  let* new_lhs = linearize_var lhs in
+  let+ new_values =
+    Phi.values phi |> Seq.to_list |>
+    Linear.List.map ~f:(fun (tid, e) ->
+        let+ e = linearize_exp e in
```

Comment:
I'm a bit confused here. Are the rhs variables of the phi getting the right prefix? It is possible we should just be deleting phi, since congruence information possibly subsumes it?

Reply:
We can't use the rhs variables with their prefixes as they would correspond to another block, which would violate the invariants of linear SSA. Getting rid of the phi may work though.